cmd/signimage/signimage.go

```diff
@@ -3,21 +3,141 @@ package main
 import (
 	"archive/tar"
 	"bytes"
+	"encoding/json"
 	"flag"
 	"fmt"
-	"github.com/docker/cli/cli/config"
-	dockertypes "github.com/docker/cli/cli/config/types"
 	"github.com/go-logr/logr"
 	"github.com/google/go-containerregistry/pkg/authn"
 	"github.com/kubernetes-sigs/kernel-module-management/internal/registry"
 	"io"
+	"io/fs"
 	"k8s.io/klog/v2/klogr"
 	"os"
 	"os/exec"
 	"path/filepath"
 	"strings"
 )
 
+// configDir should contain all the secrets available as individual files named for their keys
+// so we need to search through for the secrets we need for our pull and push repos.
+// and if we find nothingi default anonymous auth
+// we're (trying to be) tolerant in our inputs, precise in our outputs, and opaque in our comments. Its the UNIX way!
+type repoAuth struct {
+	pullRepo  string
+	pushRepo  string
+	PullAuth  authn.Authenticator
+	PushAuth  authn.Authenticator
+	configDir string
+}
+
+func NewRepoAuth(configDir string, pullRepo string, pushRepo string) *repoAuth {
+	if pushRepo == "" {
+		pushRepo = pullRepo
+	}
+	r := &repoAuth{
+		pullRepo:  pullRepo,
+		pushRepo:  pushRepo,
+		PullAuth:  nil,
+		PushAuth:  nil,
+		configDir: filepath.Clean(configDir),
+	}
+	r.PopulateAuthFromFileList()
+	if r.PullAuth == nil {
+		r.PullAuth = authn.Anonymous
+		logger.Info("no secrets for pull found, default to Anonymous")
+	}
+	if r.PushAuth == nil {
+		r.PushAuth = authn.Anonymous
+		logger.Info("no secrets for push found, default to Anonymous")
+	}
+	return r
+}
+
+func (r *repoAuth) PopulateAuthFromFileList() {
+	// not giving any secrets should short cuircuit the whole process
+	if r.configDir == "" {
+		return
+	}
+	logger.Info("walking the tree looking for secrets", "dir", r.configDir)
+	// otherwise lets hunt for repo secrets!
+	err := filepath.WalkDir(r.configDir, r.authDirFileHandler)
+	if err != nil {
+		logger.Info("no secret found, default to Anonymous", "error", err)
+		// we could return the error and die, but we want to be tolerant of secrets in the wrong format etc
+		return
+	}
+}
+
+func (r *repoAuth) authDirFileHandler(path string, d fs.DirEntry, err error) error {
+	if path == r.configDir {
+		return nil
+	}
+	if strings.Contains(path, "..") {
```

Comment:
Do we need special handling for that case? I don't think `Walk` will go back to any parent dir.
What do you think about skipping all directories anyway?

Reply:
mounting a secret as a volume doesn't just give you the keys but also gives you some extra directories called things like "..data" , they all have ".." in their names and don't contain anything we care about so that test is to make sure we skip over them.

(its Contains() rather than HasPrefix() or a slice because path includes the full directory path so "mydir/..data" )